### DISCUSSION TITLE
Building CVC5 on Linux for Windows: Dependencies unclear?

### DISCUSSION THREAD

**Discussion by kfriedberger** [cvc5/cvc5#9677] [Q&A] [answered]
Dear team of CVC5,
we (from [JavaSMT](https://github.com/sosy-lab/java-smt)) want to also include CVC5 for Windows (64bit only).
Our development platform is an Ubuntu system (e.g., Ubuntu 20.04 or 22.04), where we already build CVC5 for Linux.
We currently use version 1.0.5 of CVC5, but we could also switch to any newer version, if required (e.g., if there would be any improvements or fixes).

Based on the [existing documentation](https://cvc5.github.io/docs/cvc5-1.0.2/installation/installation.html#cross-compiling-for-windows), I am unable to cross-compile CVC5 on Linux for Windows.
Can you give more information about the following points?

### - Which dependencies are required for building CVC5 on Linux for Windows, except MinGW?
We want to use the option `--auto-download` from your build-script.
Do we need to provide any other software or content, such as GMP/libgmp or Java/JDK?

### - How does CVC5 solve the issue of the thread-library that is not available for MinGW? It seems to be included in the `time-limits` component and causes the following error:
```
cvc5/src/main/time_limit.cpp: In function ‘cvc5::main::TimeLimit cvc5::main::install_time_limit(uint64_t)’:
cvc5/src/main/time_limit.cpp:116:8: error: ‘thread’ is not a member of ‘std’
```
I tried a solution based on https://github.com/meganz/mingw-std-threads, which seems to avoid the mentioned error. However, I have no idea whether this would produce correct library in the final processing (currently there are other problems, so the building aborts before finishing).

### - How does the compilation process build GMP/libgmp and link against it?
  The final linking process aborts with a long list of undefined references:
```
[100%] Built target main
Consolidate compiler generated dependencies of target main-test
[100%] Built target main-test
Consolidate compiler generated dependencies of target cvc5-bin
[100%] Linking CXX executable ../../bin/cvc5.exe
/usr/bin/x86_64-w64-mingw32-ld: ../libcvc5.a(cvc5.cpp.obj):cvc5.cpp:(.text+0xf5e8): undefined reference to `__gmpz_init_set'
/usr/bin/x86_64-w64-mingw32-ld: ../libcvc5.a(cvc5.cpp.obj):cvc5.cpp:(.text+0xf64e): undefined reference to `__gmpz_cmp_ui'
/usr/bin/x86_64-w64-mingw32-ld: ../libcvc5.a(cvc5.cpp.obj):cvc5.cpp:(.text+0x1c87f): undefined reference to `__gmpz_cmp_ui'
... and several other undefined references
```

### - How does the compilation process include JNI? Which parameters or environment variables need to be set, such that `./configure --java-bindings ...` finds JNI for Windows. The issue is as follows:
```
-- Found Java: /usr/bin/java (found version "17.0.6") found components: Development 
CMake Error at /usr/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find JNI (missing: JAVA_AWT_LIBRARY JAVA_JVM_LIBRARY
  JAVA_INCLUDE_PATH JAVA_INCLUDE_PATH2 JAVA_AWT_INCLUDE_PATH)
Call Stack (most recent call first):
  /usr/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.22/Modules/FindJNI.cmake:382 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  src/api/java/CMakeLists.txt:153 (find_package)
```
  Do we need to provide any directory containing the Windows-sources of JNI or JDK?
  Which of those missing environment variables are really required?
  Which version of Java is required? The example from above is using Java17, for testing. Normally, we are using Java11 for building the Linux-library, so it should also be working for Windows, right?

Best,
Karlheinz

**Comment by daniel-larraz** [comment, marked as answer]
@kfriedberger Sorry for the late reply. Please find my response below in case you are still interested.

We currently offer cross-compilation of cvc5 binaries and libraries on Linux for Windows, but not for the Java bindings. The only supported method of building the Java bindings for Windows is by building the cvc5 libraries and the Java bindings directly on Windows. Please see the [INSTALL.rst](https://github.com/cvc5/cvc5/blob/main/INSTALL.rst) document for up-to-date instructions on building cvc5 on Windows.

In addition, we now offer pre-compiled cvc5 binaries, libraries, and Java bindings for Windows on the release page.

> Which dependencies are required for building cvc5 on Linux for Windows, except MinGW?

Cross-compilation of the cvc5 libraries for Windows using the  `--auto-download` option does not require any additional dependencies other than the ones for building them on Linux (except MinGW).

> How does cvc5 solve the issue of the thread-library that is not available for MinGW?

Cross-compiling cvc5 for Windows requires the POSIX version of MinGW-w64. On some Linux distributions, this is the default variant. On others, like Ubuntu, you may need to set it manually as follows:

```
sudo update-alternatives --set x86_64-w64-mingw32-gcc /usr/bin/x86_64-w64-mingw32-gcc-posix
sudo update-alternatives --set x86_64-w64-mingw32-g++ /usr/bin/x86_64-w64-mingw32-g++-posix
```

I updated the INSTALL.rst document to clarify this.

> How does the compilation process build GMP/libgmp and link against it?

When the `--auto-download` option is provided, the build system takes care of downloading and building GMP for you. The issue with undefined references should be fixed in v1.2.0.

> How does the compilation process include JNI?

Using `./configure --java-bindings` should be enough for building the Java bindings on Windows, provided the JDK is installed on your system. As mentioned above, the build system is not yet ready for cross-compiling the Java bindings on Linux; it only supports the binaries and C++ libraries. The error message you encountered could occur if multiple versions of Java were installed on your system. PR #11163, which was just merged, fixes this issue. Alternatively, you can specify a Java installation prefix explicitly by passing `-DJAVA_HOME=/path/to/java/home` to `configure.sh`.

Please let me know if you have any other questions.

Thanks,
Daniel

**Reply by kfriedberger** [reply]
Thank you for your response. Currently, the topic is no longer urgent. We will follow up on the matter if necessary.
